loader/0032-loader-plugins.md

```diff
@@ -0,0 +1,328 @@
+
+/**
+* Invoked when tentative LoadOptions are removed at the end of a cycle.
+*/
+public abstract void onLoadOptionRemoved(LoadOption option);
+
+/**
+* Called at the start of each cycle to encode this rule in sat4j.
+*/
+public abstract void define(IRuleDefiner definer);
+}
+```
+</details>
+
+
+# Drawbacks
+
+All existing behavior can be implemented using loader plugins. As such, the largest
+drawback may be the increased scope for the Quilt project as a whole, as we will no
+longer be limited to mods specifically. This is less of a drawback and more of an
+organizational change that needs to be considered moving forward.
+
+
+# Rationale and Alternatives
+
+While we could continue with the current status quo of loading mods from the mods
+folder and only the mods folder, this is a rigid and outdated way of doing so that
+does not allow for iteration and innovation.
+
+If loader plugins are not implemented, we would have a very hard time
+differentiating ourselves from other mod loaders. Plugins will allow both us and
+the community itself to add features to Quilt loader easily, making it much easier
+for innovation in where mods come from and how they are loaded.
+
+
+# Prior Art
+
+If this has been done before by some other project, explain it here. This could
+be positive or negative. Discuss what worked well for them and what didn't.
+
+Similar projects have been looked at by others in the past:
+- FabLabs very briefly explored this concept [here](https://github.com/FabLabsMC/fabric-loader/tree/feature/modproviders).
+- [@Chocohead](https://github.com/Chocohead) has implemented [loading extra mods alongside Fabric Loader](https://github.com/Chocohead/Modjam/blob/master/src/com/chocohead/sm/loader/PreLoader.java), but not as actual Fabric mods.
+
+None of them have been implemented into the mod loader itself, and as such this is
+the first project to allow such liberal changes to the mod loading process.
+
+
+# Unresolved Questions
+
+Before loader plugins can be considered fully implemented, CHASM will need to be
+functional so that all bytecode modifications can go through a single source of
+truth.
+
+For the purposes of this RFC, I did not go into the drawbacks of
+dependency downloading specifically, as this RFC is interested more generally
+in the idea of loader plugins as a whole. How dependency downloading is implemented
+and other concerns regarding that will take place at a later date, in another RFC.
+
+
+# Expected Response
+
+Loader plugins have been one of the earliest features Quilt promised to deliver. As
+such, I expect the community will largely be excited to see this feature
+implemented. Projects such as [Bundle Installer](https://github.com/FoundationGames/Bundle-Installer) will be able to make use of loader plugins right from day one, and
+I expect that we'll see a lot of similar projects that aim to make the mod
+installation process as easy as possible for end users.
```

Comment:
You already have these now - they're called launchers! In fact we had them even with Modloader, people having trouble managing their mods is nothing new, nor are many of the solutions to trying to fix said troubles. I'd be sceptical of why having multiple similar but different systems is making the process easier than either providing a single uniform, centralised one or just pushing for better launcher integration in the first place.